extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -957,7 +961,8 @@ private List<SegmentIdWithShardSpec> generateSegmentIdsWithShardSpecs(
       final RowSignature signature,
       final ClusterBy clusterBy,
       final ClusterByPartitions partitionBoundaries,
-      final boolean mayHaveMultiValuedClusterByFields
+      final boolean mayHaveMultiValuedClusterByFields,
+      final Boolean isStageOutputEmpty
   ) throws IOException
   {
     if (destination.isReplaceTimeChunks()) {
```

Comment:
it seems to me that both methods called from here do early-return based on `isStageOutputEmpty` 
why not handle `isStageOutputEmpty` here or at the callsite of `generateSegmentIdsWithShardSpecs` in `startStages`?

Reply:
Yeah, I feel delegating what to do with an empty stage output should be the responsibility of the appropriate query type. In this case, both happen to return early, but I think as the MSQ engine evolves and/or we want different behavior, it'd be better to have that determination in these delegating functions